system/Database/BaseBuilder.php

```diff
@@ -2051,6 +2067,72 @@ public function onConstraint($set)
         return $this;
     }
 
+    /**
+     * Sets data source as a query for insert/update/upsert
+     *
+     * @param BaseBuilder|RawSql|string $query
+     */
+    public function fromQuery($query): BaseBuilder
+    {
+        if (! empty($query)) {
+            if ($query instanceof BaseBuilder) {
+                $query = $query->getCompiledSelect();
+            } elseif ($query instanceof RawSql) {
+                $query = $query->__toString();
+            }
+
+            if (is_string($query)) {
+                $this->QBOptions['fromQuery'] = $query;
+                $this->QBKeys                 = $this->db->protectIdentifiers($this->fieldsFromQuery($query));
+                $this->QBSet                  = [];
+            }
+        }
+
+        return $this;
+    }
+
+    /**
+     * Gets column names from a select query
+     */
+    protected function fieldsFromQuery(string $sql): array
```

Comment:
Parsing SQL with regular expressions is not a good way, but are you sure this way is not a problem?

Could you move this method to a separate class and write tests for it?

Reply:
I'm not sure why it should be in another class. This method could potentially be useful for other things within this class.

I can write a test for it. I wish I still had the one I used when I built the method. It had every conceivable thing in it.

Reply:
Or maybe you were meaning a separate test class?

Reply:
> I'm not sure why it should be in another class.

Because this BaseBuilder class is too big. It has more than 3,000 lines!
and the Class Complexity is 348 !!!

We should follow SOILD principles more.

`4.3` branch:
```console
$ phploc system/Database/BaseBuilder.php 
phploc 7.0.2 by Sebastian Bergmann.

Size
  Lines of Code (LOC)                             3212
  Comment Lines of Code (CLOC)                    1111 (34.59%)
  Non-Comment Lines of Code (NCLOC)               2101 (65.41%)
  Logical Lines of Code (LLOC)                     588 (18.31%)
    Classes                                        588 (100.00%)
      Average Class Length                         588
        Minimum Class Length                       588
        Maximum Class Length                       588
      Average Method Length                          4
        Minimum Method Length                        1
        Maximum Method Length                       28
      Average Methods Per Class                    129
        Minimum Methods Per Class                  129
        Maximum Methods Per Class                  129
    Functions                                        0 (0.00%)
      Average Function Length                        0
    Not in classes or functions                      0 (0.00%)

Cyclomatic Complexity
  Average Complexity per LLOC                     0.59
  Average Complexity per Class                  348.00
    Minimum Class Complexity                    348.00
    Maximum Class Complexity                    348.00
  Average Complexity per Method                   3.69
    Minimum Method Complexity                     1.00
    Maximum Method Complexity                    19.00
```

Reply:
If this continues, this class will only get longer. It is my hope that this class will not grow any larger. Because it is difficult for humans to read code that is too long.

Aside from that, if you are going to adopt the way of parsing SQL statements, please add tests for just this method.

Reply:
Well if we need to break this class down then we should do it in a methodical way and not just random methods. Maybe for 5.0 we can refactor things?

Test has been added for parsing SQL. Feel free to make suggestions of more ways to test.

Reply:
> Well if we need to break this class down then we should do it in a methodical way and not just random methods. 

Yes.

> Maybe for 5.0 we can refactor things?

Maybe it is better. We need to discuss more with other members.